cupy/core/_fusion_trace.pyx

```diff
@@ -0,0 +1,624 @@
+import numpy
+
+from cupy.core import _kernel
+from cupy.core import _reduction
+from cupy.core import core
+from cupy.core._fusion_interface import _VariableProxy
+from cupy.core._fusion_interface import _ArrayProxy
+from cupy.core import _fusion_thread_local
+from cupy.core import _fusion_variable
+from cupy.core._fusion_variable import _AbstractDim
+from cupy.core._fusion_variable import _TraceScalar
+from cupy.core._fusion_variable import _TraceArray
+from cupy.core._fusion_variable import _VariableSet
+from cupy.core import _fusion_op
+from cupy.core import _fusion_optimization
+from cupy import util
+
+from cupy.cuda.function cimport Arg
+
+
+_thread_local = _fusion_thread_local.thread_local
+_accepted_types = (int, float, bool, complex, numpy.generic)
+
+
+cdef class _ShapeConstraints:
+    """A data structure that manages the conditions between the shapes.
+    """
```

Comment:
A high-level description on how are shape constraints compared would be helpful!